colossalai/inference/core/request_handler.py

```diff
@@ -1,48 +1,210 @@
+        return len(self.prefill) / len(self.decoding) >= self.ratio
+
+    def is_empty(self):
+        return not self.decoding and not self.prefill
+
 
 class RequestHandler:
     """
     RequestHandler is the core for handling existing requests and updating current batch.
     During generation process, we call schedule function each iteration to update current batch.
 
     Args:
-        cache_config: Configuration for initialize and manage kv cache.
+       inference_config: Configuration for initialize and manage kv cache.
     """
 
-    def __init__(self, cache_config) -> None:
-        self.cache_config = cache_config
-        self._init_cache()
-        self.waiting_list: List["Reqseq"] = []
-        self.running_list: List["Reqseq"] = []
+    def __init__(self, inference_config: InferenceConfig, model_config: PretrainedConfig) -> None:
+        self.inference_config = inference_config
+        self._init_cache(model_config)
 
-    def _init_cache(self):
-        """
-        Initialize the cache manager with cache config.
-        """
+        self.running_list: RunningList = RunningList(inference_config.prefill_ratio)
+        self.waiting_list: List[List] = [[], [], []]
+        self.done_list: List[Sequence] = []
+        self.running_batch = BatchInfo(is_prompts=False)
+        self.prefill_batch = BatchInfo(is_prompts=True)
+
+    def _init_cache(self, model_config):
+        self.cache_manager = KVCacheManager(self.inference_config, model_config)
+
+    def _has_waiting(self) -> bool:
+        return any(lst for lst in self.waiting_list)
 
     def schedule(self):
         """
         The main logic of request handler.
         """
+        if self._has_waiting():
+            # Try to allocate cache blocks for the sequence using a priority of prompt length.
+            for lst in reversed(self.waiting_list):
+                if lst:
+                    seq = lst[0]
+                    if seq.prompt_len > self.inference_config.max_input_len:
+                        # If the prompt length is longer than max_input_len, abort the sequence.
+                        self.abort_sequence(seq.request_id)
+                        break
+                    # Try to allocate cache blocks for the sequence.
+                    if self.cache_manager.num_available_blocks > self.cache_manager.max_blocks_per_sequence:
+                        # If succeed, add the sequence to running list.
+                        self.running_list.append(seq)
+                        self.cache_manager.allocate_context_from_block_table(seq.block_table_index, seq.prompt_len)
+                        lst.pop(0)
 
-    def add_sequence(self, reqseq: "Reqseq"):
+        if self.running_list.ready_for_prefill():
+            for seq in self.running_list.prefill:
+                seq.mark_running()
+            self.prefill_batch.init_batch(self.running_list.prefill)
+            return self.prefill_batch
+
+        return self.running_batch
```

Comment:
Does schedule return without processing the running_batch?

Reply:
If we have pre-allocated all needed space, we don't need to change running batch as all the reqs are updated already.